File: packages/python/src/daraja/utils/tracing.py
from typing import Any, Callable, Optional

from opentelemetry import trace
from opentelemetry.trace import Span as OTelSpan, StatusCode


class Span:
    def set_attribute(self, key: str, value: Any) -> None:
        pass

    def add_event(self, name: str, attributes: Optional[dict[str, Any]] = None) -> None:
        pass

    def record_exception(self, error: Exception) -> None:
        pass

    def set_status(self, code: str, message: Optional[str] = None) -> None:
        pass

    def end(self) -> None:
        pass


class Tracer:
    def start_span(self, name: str, attributes: Optional[dict[str, Any]] = None) -> Span:
        return Span()

    def with_span(self, name: str, attributes: Optional[dict[str, Any]] = None) -> "SpanContext":
        return SpanContext(self, name, attributes)


class SpanContext:
    def __init__(self, tracer: Tracer, name: str, attributes: Optional[dict[str, Any]] = None) -> None:
        self._tracer = tracer
        self._name = name
        self._attributes = attributes

    def __enter__(self) -> Span:
        self._span = self._tracer.start_span(self._name, self._attributes)
        return self._span

    def __exit__(self, exc_type: Any, exc_val: Any, exc_tb: Any) -> None:
        if exc_val is not None:
            self._span.record_exception(exc_val)
            self._span.set_status("error", str(exc_val))
        self._span.end()


class NoopTracer(Tracer):
    pass


class OTelSpanWrapper(Span):
    def __init__(self, span: OTelSpan) -> None:
        self._span = span

    def set_attribute(self, key: str, value: Any) -> None:
        self._span.set_attribute(key, value)

    def add_event(self, name: str, attributes: Optional[dict[str, Any]] = None) -> None:
        self._span.add_event(name, attributes)

    def record_exception(self, error: Exception) -> None:
        self._span.record_exception(error)

    def set_status(self, code: str, message: Optional[str] = None) -> None:
        if code == "error":
            self._span.set_status(StatusCode.ERROR, message)
        else:
            self._span.set_status(StatusCode.OK)

    def end(self) -> None:
        self._span.end()


class OpenTelemetryTracer(Tracer):
    def __init__(self, name: str = "mpesa-sdk", version: str = "0.2.0") -> None:
        self._tracer = trace.get_tracer(name, version)

    def start_span(self, name: str, attributes: Optional[dict[str, Any]] = None) -> Span:
        span = self._tracer.start_span(name, attributes=attributes)
        return OTelSpanWrapper(span)


def create_tracer(logger: Any = None) -> Tracer:
    return OpenTelemetryTracer()


def with_span(tracer: Tracer, name: str, attributes: Optional[dict[str, Any]] = None) -> SpanContext:
    return tracer.with_span(name, attributes)

File: sdks/typescript/src/utils/tracing.ts
import { trace, type Attributes, Span as OTelSpan, SpanStatusCode } from "@opentelemetry/api";
import type { Logger, TelemetrySpan, Tracer } from "../types/index.js";

export class NoopSpan implements TelemetrySpan {
  end(): void {}
  setAttribute(_key: string, _value: string | number | boolean): void {}
  addEvent(_name: string, _attributes?: Record<string, unknown>): void {}
  recordException(_error: unknown): void {}
  setStatus(_code: string, _message?: string): void {}
}

export class NoopTracer implements Tracer {
  startSpan(): TelemetrySpan {
    return new NoopSpan();
  }
}

class OTelTelemetrySpan implements TelemetrySpan {
  constructor(private readonly span: OTelSpan) {}

  end(): void {
    this.span.end();
  }

  setAttribute(key: string, value: string | number | boolean): void {
    this.span.setAttribute(key, value);
  }

  addEvent(name: string, attributes?: Record<string, unknown>): void {
    this.span.addEvent(name, attributes as Attributes);
  }

  recordException(error: unknown): void {
    this.span.recordException(error as { message: string });
  }

  setStatus(code: string, message?: string): void {
    if (code === "error") {
      this.span.setStatus({ code: SpanStatusCode.ERROR, message });
    } else {
      this.span.setStatus({ code: SpanStatusCode.OK });
    }
  }
}

export class OpenTelemetryTracer implements Tracer {
  private readonly tracer: ReturnType<typeof trace.getTracer>;

  constructor(name = "mpesa-sdk", version = "0.2.0") {
    this.tracer = trace.getTracer(name, version);
  }

  startSpan(name: string, attributes?: Record<string, string>): TelemetrySpan {
    const span = this.tracer.startSpan(name, {
      attributes: attributes as Record<string, string>,
    });
    return new OTelTelemetrySpan(span);
  }
}

export function createTracer(_logger?: Logger): Tracer {
  return new OpenTelemetryTracer();
}

export async function withSpan<T>(
  tracer: Tracer,
  name: string,
  fn: () => Promise<T>,
  attributes?: Record<string, string>,
): Promise<T> {
  const span = tracer.startSpan(name, attributes);
  try {
    const result = await fn();
    return result;
  } catch (error) {
    span.recordException(error);
    span.setStatus("error", error instanceof Error ? error.message : String(error));
    throw error;
  } finally {
    span.end();
  }
}

export async function withSpanSync<T>(
  tracer: Tracer,
  name: string,
  fn: () => T,
  attributes?: Record<string, string>,
): Promise<T> {
  const span = tracer.startSpan(name, attributes);
  try {
    return fn();
  } catch (error) {
    span.recordException(error);
    span.setStatus("error", error instanceof Error ? error.message : String(error));
    throw error;
  } finally {
    span.end();
  }
}
